pkg/status/reporter.go

```diff
@@ -0,0 +1,248 @@
+		r.logCtx.Infof("Node status update interval updated")
+		r.ticker = time.NewTicker(time.Duration(interval) * time.Second)
+	}
+}
+
+// Cleanup resources owned by this reporter.
+func (r *reporter) cleanup() {
+	r.ticker.Stop()
+}
+
+// KillAndWait sends done signal to reporter goroutine and wait until the
+// goroutine of the reporter terminated.
+func (r *reporter) KillAndWait() {
+	r.done <- struct{}{}
+	<-r.term
+	r.logCtx.Debug("Node status reporter terminated.")
+}
+
+// Called when the caller needs to send a new version of request.
+func (r *reporter) RequestUpdate(request *apiv3.CalicoNodeStatus) {
+	r.ch <- request
+}
+
+// Return if the current status of the reporter has the same spec with
+// the status passed in.
+func (r *reporter) HasSameSpec(status *apiv3.CalicoNodeStatus) bool {
+	return reflect.DeepEqual(r.status.Spec, status.Spec)
+}
+
+// ReportStatus call reportStatus function.
+// ToDo error handling and update conditions.
+func (r *reporter) ReportStatus() {
+	_ = r.reportStatus()
+}
+
+// run is the main reporting loop, it loops until done.
+func (r *reporter) run() {
+	r.logCtx.Debug("Start new goroutine to report node status")
+
+	for {
+		select {
+		case latest := <-r.ch:
+			// Received an update of node status resource.
+			if latest.Name != r.name {
+				r.logCtx.Errorf("node status reporter receive request with different name (%s), ignore it", latest.Name)
+				break
+			}
+
+			r.status = latest
+			r.checkAndUpdateTicker(latest.Spec.UpdatePeriodSeconds)
+			// kick start node status update
+			r.ReportStatus()
+
+		case <-r.ticker.C:
+			// Todo check resource and update condition.
+			r.ReportStatus()
+
+		case <-r.done:
+			r.cleanup()
+			r.term <- struct{}{}
+			return
+		}
+	}
+}
+
+// reportStatus queries Bird or other components and update node status resource.
+func (r *reporter) reportStatus() error {
+	// The idea here is that we either update everything successfully or we update nothing.
+
+	// Make a local copy first.
+	status := *r.status
+
+	for _, ipv := range []populator.IPFamily{populator.IPFamilyV4, populator.IPFamilyV6} {
+		// Populate status from registered populators.
+		for _, class := range r.status.Spec.Classes {
+			p, ok := r.populators[ipv][class]
+			if !ok {
+				r.logCtx.Warningf("Wrong class (%s) requested for node status reporter", class)
+				continue
+			}
+			err := p.Populate(&status)
+			if err != nil {
+				// If we hit any error, stop the entire update process.
+				r.logCtx.WithError(err).Errorf("failed to populate status for ipv%s class %s", string(ipv), string(class))
+				return err
+			}
+		}
+	}
+
+	r.logCtx.Debug("Status updated by populators")
+
+	if reflect.DeepEqual(status.Status, r.status.Status) {
+		// Nothing has changes since last time we updated.
+		return nil
+	}
+
+	var err error
+	var updatedResource *apiv3.CalicoNodeStatus
+	// Update resource
+	for i := 0; i < 3; i++ {
+		ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
```

Comment:
I suppose this has a problem if the update interval set by the user is smaller than our timeout!

We might want to consider not retrying if the update interval is suitably small - say less than 30 seconds - since we're going to retry again shortly anyway.

Reply:
Good point!